tools/oesign/oesign.c

```diff
@@ -671,3 +793,76 @@ int oedigest(const char* enclave, const char* conffile, const char* digest_file)
 done:
     return ret;
 }
+
+char hexchar2int(char ch)
+{
+    if (ch >= '0' && ch <= '9')
+        return ch - '0';
+    if (ch >= 'a' && ch <= 'f')
+        return 10 + ch - 'a';
+    if (ch >= 'A' && ch <= 'F')
+        return 10 + ch - 'A';
+    return 0;
+}
+
+ bool is_hex(char ch)
+{
+    return (ch >= '0' && ch <= '9') || (ch >= 'a' && ch <= 'f') ||
+           (ch >= 'A' && ch <= 'F');
+}
+
+unsigned char hexpair2char(char a, char b)
+{
+    return (unsigned char)((hexchar2int(a) << 4) | hexchar2int(b));
+}
+
+int uuid_from_string(str_t* str, uint8_t* uuid, size_t expected_size)
+{
+    size_t index = 0;
+    size_t size = 0;
+    char* id_copy;
+    char value = 0;
+    bool firstDigit = true;
+
+    id_copy = oe_strdup(str_ptr(str));
+    if (!id_copy)
+        return -1;
+
+    size = strlen(id_copy);
+    if (size != 36)
+    {
+        free(id_copy);
+        return -1;
+    }
+
+    index = 0;
+
+    for (size_t i = 0; i < size; ++i)
+    {
+        if (id_copy[i] == '-')
+            continue;
+
+        if (index >= expected_size || !is_hex(id_copy[i]))
+        {
+            free(id_copy);
+            return -1;
+        }
+
+        if (firstDigit)
+        {
+            value = id_copy[i];
+            firstDigit = false;
+        }
+        else
+        {
+            uuid[index++] = hexpair2char(value, id_copy[i]);
+            firstDigit = true;
+        }
+    }
+    if (index < expected_size)
+    {
+        free(id_copy);
+        return -1;
+    }
+    return 0;
```

Comment:
id_copy is a local copy of str and is never passed back to caller. It needs to be freed regardless of success or failure.